pkg/controller/resolvedofferings/volumeclass.go

```diff
@@ -0,0 +1,112 @@
+package resolvedofferings
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/acorn-io/baaah/pkg/typed"
+	v1 "github.com/acorn-io/runtime/pkg/apis/api.acorn.io/v1"
+	internalv1 "github.com/acorn-io/runtime/pkg/apis/internal.acorn.io/v1"
+	"github.com/acorn-io/runtime/pkg/config"
+	"github.com/acorn-io/runtime/pkg/volume"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func resolveVolumeClasses(ctx context.Context, c kclient.Client, app *internalv1.AppInstance) error {
+	if len(app.Status.AppSpec.Volumes) == 0 {
+		return nil
+	}
+
+	volumeClasses, defaultVolumeClass, err := volume.GetVolumeClassInstances(ctx, c, app.Namespace)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range typed.Sorted(volumeClasses) {
+		vc := entry.Value
+		if vc.Default && vc.Name != defaultVolumeClass.Name {
+			return fmt.Errorf("cannot resolve volume classes because two defaults volume classes exist: %s and %s", defaultVolumeClass.Name, vc.Name)
+		}
+	}
+
+	if app.Status.ResolvedOfferings.Volumes == nil {
+		app.Status.ResolvedOfferings.Volumes = make(map[string]internalv1.VolumeResolvedOffering)
+	}
+
+	volumeBindings := volume.SliceToMap(app.Spec.Volumes, func(vb internalv1.VolumeBinding) string {
+		return vb.Target
+	})
+
+	for name, vol := range app.Status.AppSpec.Volumes {
+		resolvedVolume := app.Status.ResolvedOfferings.Volumes[name]
+		vol = volume.ResolveVolumeRequest(vol, volumeBindings[name], resolvedVolume)
+
+		// This is a bit of a hack as we're migrating away from the VolumeSize field. Essentially,
+		// we want to ensure that app.Status.Volumes[name] always has a size set. If the VolumeSize
+		// field has been set in the past, we want to migrate that over to be set on app.Status.Volumes[name].
+		// There is another edge case where the Size field was set by a VolumeClass's default size. In this
+		// case we want to leave the Size field alone.
+		if app.Status.ResolvedOfferings.VolumeSize != nil && resolvedVolume.Size == "" {
+			resolvedVolume.Size = internalv1.Quantity(app.Status.ResolvedOfferings.VolumeSize.String())
+		}
+
+		bound := volumeBindings[name].Volume != ""
+
+		if _, alreadySet := app.Status.ResolvedOfferings.Volumes[name]; !alreadySet && !bound {
```

Comment:
This makes it seem like it is not possible to change a volume's settings. For example, if a volume as been resolved and is `alreadySet`, then would it be possible to change its size or access modes?

The difference of what this did before was that it was only calculating the defaults, and we decided that the defaults for a volume shouldn't change because it causes too many issues. However, now this is the completely "resolved" volume information, which should be allowed to change if the user tries to change it via an Acornfile update (I think your logic allows the user to change volume information via the command line).

Reply:
Oh, good point. I will work on fixing this.

Reply:
So, from what I understand, you cannot change the access modes of an existing PVC in Kubernetes, but you can increase its size. I heavily modified this function so that the size can be changed even if the volume offering has been previously resolved, but the volume class and access modes cannot be changed after the volume is created. Let me know if it looks good to you. From my testing, everything seems to me to be working properly.

Reply:
It doesn't seem to set a different size if a volume is `alreadySet` and not `bound`.